[AC] Backward Pass Aware AC - adding hooks to partitioner to pass callable

paryxyt opened the pull request:
Summary: same as title. Plan is to pass a callable to the partitioner to perform custom autoAC via an ILP. This is the same as a previous diff D63714905 which was landed and then subsequently reverted by PyTorch Release Engineering because of a failing unit test (https://hud.pytorch.org/pytorch/pytorch/commit/f7b8d36c287867f94c1848cccf104eebc3f0e98d?fbclid=IwY2xjawFz25lleHRuA2FlbQIxMQABHfbEcGQ7BC9k32zz5TyPN_UsuiwrR-wr40R1KOto8ie5MLGPkzvj2Q-R8Q_aem_kDWkwz9VReSsnzLyvwTMBA). We think the unit test is buggy, and we also fix the same.

Test Plan: tbd

huydhn commented:
@pytorchbot drci

basilwong reviewed (approved):
Spoke offline, got approval from compiler team 👍

paryxyt committed:
[AC] Backward Pass Aware AC - adding hooks to partitioner to pass callable (#137785)

Summary:

same as title. Plan is to pass a callable to the partitioner to perform custom autoAC via an ILP. This is the same as a previous diff D63714905 which was landed and then subsequently reverted by PyTorch Release Engineering because of a failing unit test (https://hud.pytorch.org/pytorch/pytorch/commit/f7b8d36c287867f94c1848cccf104eebc3f0e98d?fbclid=IwY2xjawFz25lleHRuA2FlbQIxMQABHfbEcGQ7BC9k32zz5TyPN_UsuiwrR-wr40R1KOto8ie5MLGPkzvj2Q-R8Q_aem_kDWkwz9VReSsnzLyvwTMBA). We think the unit test is buggy, and we also fix the same.

Test Plan: tbd

Reviewed By: basilwong

Differential Revision: D64246495

ezyang commented:
@pytorchbot revert -c ignoredsignal -m "breaks lint"

huydhn commented:
@pytorchbot rebase

paryxyt committed:
Update partitioners.py - fixing lint error

basilwong commented on torch/_functorch/partitioners.py:
```diff
@@ -1680,18 +1690,28 @@ def get_recomputable_banned_nodes(banned_nodes: List[fx.Node]) -> List[fx.Node]:
+            # if idx in all_recomputable_banned_nodes:
+            try:
+                dont_ban.add(all_recomputable_banned_nodes[idx])
+            except BaseException:
```
What is the expected exception here? Are we expected an index error?

I am a bit surprised you can't use:

```
            # if idx in all_recomputable_banned_nodes:
```

paryxyt replied:
@basilwong - we can use that, but I think it adds an O(n) operation inside the larger for loop. This operation actually broke one of the pre-existing unit tests, because the flop count increased ....

paryxyt replied:
(The expected exception here is that idx is not in all_recomputable_banned_nodes)

huydhn committed:
Update partitioners.py to fix another lint warnings

huydhn commented:
@pytorchbot merge -f 'Lint jobs have passed, the rest will stay the same'
